### ISSUE TITLE
Clarification: boolean/shift ops and V (signed overflow)

### ISSUE THREAD

**Issue by leonardt** [StanfordAHA/CGRAGenerator#62]
This isn't specified in the google doc, what is the expected/desired behavior for the `V` flag (signed overflow) and the boolean operations `and`, `or`, and `xor`? Based on my tests, it looks like the boolean ops clear the flag (always 0), which is a typical behavior in modern processors, but just wanted to make this decision explicit in terms of specification.

Also, it seems that the shift ops also clear the overflow flag.

Also, per https://github.com/StanfordAHA/CGRAGenerator/issues/61, do we generally have any expectations for the `V` flag in unsigned mode? Or will it be undefined.

**Comment by kong0329** [comment]
| C | V (signed) | V (unsigned)
-- | -- | -- | --
Default | (a + b) >= 2**16 | 0 | 0
add | (a + b + d) >= 2**16 | (a[15]==b[15]) and (a[15] != (a+b+d)[15]) | (a + b + d) >= 2**16
sub | (a + ~b + 1) >= 2**16 | (a[15]!=[b15]) and (a[15] != (a+~b+1)[15]) | 0
abs | (~a + 1) >= 2**16 | (~a + 1)[15] | (~a + 1)[15]
gte | (a + ~b + 1) >= 2**16 | default | default
lte | (a + ~b + 1) >= 2**16 | default | default
mul0 | default | If (a[15]==b[15])  (a*b)[15] | (a*b)[31:16] == 0
Else (a*b)[15] == 0 and (a!=0 or b!=0)
mul1 | default | If (a[15]==b[15])  (a*b)[15] | (a*b)[31:24] == 0
Else (a*b)[15] == 0 and (a!=0 or b!=0)
mul2 | default | 0 | 0
others | default | default | default



This is how condition flags are implemented in RTL.
This is just implementation, not spec, and I have to get confirmed whether this is a correct SPEC.

**Comment by leonardt** [comment]
@kong0329 thanks for listing that out, I will keep this issue open until we finalize the spec. Then we can test the implementation and make sure it checks out.

**Comment by kong0329** [comment]
@leonardt No problem. That's good!

**Comment by leonardt** [comment]
I'm inclined to say that the expected behavior of the V flag (signed overflow) should be undefined when in unsigned mode. This leaves it up to the implementation to decide (e.g. if there's a better choice for optimization reasons).

Is there a case when the V flag is useful in unsigned mode?

**Comment by leonardt** [comment]
Similar reasoning could be used to argue that the semantics of the C flag should be undefined for signed mode.

**Comment by steveri** [comment]
> the expected behavior of the V flag (signed overflow) should be undefined when in unsigned mode.

Seems like unsigned V should be the same as C, i.e. `if {unsigned} then {V===C}`


> Similar reasoning could be used to argue that the semantics of the C flag should be undefined for signed mode.

I'd say it differently: the behavior of C flag should be the same for both signed and unsigned mode

Also: how are these (C and V) defined in ARM?  That should probably be the default.  I'm a big fan of consistency...

**Comment by phanrahan** [comment]
We should use the standard definition of V.

V = (msb1==msb22)&(msb1!=msb)

where msb1 and msb2 are the sign bits of the inputs and msb is the sign bit of the output.

Basically, this says that if adding two numbers of the same sign, the result should have the
same sign; otherwise, an overflow has occurred.

There are lots of ways of computing this, but they should yield the same result.

Signed vs unsigned should probably only affect  lte and gte ops. Not add and sub. It can also
effect extended precision arithmetic if sign extension may be necessary.

The power of 2’s complement is that signed and unsigned arithmetic are usually the same. Let’s only
change that if necessary.

Sorry for being late to the game. I haven’t been following all this … 

Pat

**Comment by leonardt** [comment]
I'm okay with using the standard definition of `V` that Pat described, that seems simplest and makes sense to me.

One thing to clarify is that you say "when adding two numbers of the same sign, the result should have the same sign...". 

For the case of multiplication, should we still have `V` correspond to signed overflow in the add? Or have it as signed overflow in the multiplication operation. 

The latter would update your wording to be "when the two inputs have the same sign, the result should have the same sign or an overflow has occured". This corresponds to the current demonstrated behavior with `abs` (abs changes the sign if the input is negative, which then sets the `V` flag).

**Comment by leonardt** [comment]
Also, based on your wording you say that sign/unsigned should only affect lte/gte ops, so C and V should demonstrate the same behavior in signed/unsigned mode? That makes sense to me, just wanted to clarify

**Comment by tema8** [comment]
For ABS(a), V will be set only if a = 'h8000

**Comment by cs107e-student-0** [comment]
Thanks Artem, correction: `abs` does not follow the model I described. 

It seems that what Artem describes and Taeyoung describes in the table above is mismatched. The table says `(~a + 1)[15]` should set the `V` flag if the result of the two's complement negation is negative. Artem says it will only set it if a = `0x8000`, which I assume is because if we take the two's complement negation, we flip the bits and add one, which gives us a negative number from the negation of a negative number. So I think what Artem is describing aligns with what Pat is describing: if the result of the negation of a negative number is negative, there is a signed overflow.

**Comment by leonardt** [comment]
Woops, commented on the wrong account! (testing cs107e stuff hehe), original comment appended here under my actual account.

Thanks Artem, correction: `abs` does not follow the model I described. 

It seems that what Artem describes and Taeyoung describes in the table above is mismatched. The table says `(~a + 1)[15]` should set the `V` flag if the result of the two's complement negation is negative. Artem says it will only set it if a = `0x8000`, which I assume is because if we take the two's complement negation, we flip the bits and add one, which gives us a negative number from the negation of a negative number. So I think what Artem is describing aligns with what Pat is describing: if the result of the negation of a negative number is negative, there is a signed overflow.

**Comment by kong0329** [comment]
@tema8 Original code was not implemented in that way. 
```
    PE_ABS_OP: begin
        add_a[0]    = ~op_a;
        add_b[0]    = 0;
        add_c_in[0] = 1'b1;
        res_p_w     = op_a[DATA_MSB];
        ovfl        = res_w[DATA_MSB];
        res_w       = diff_sign ? op_a : add_res[0];
    end
```
Since it is blocking statement, in order to follow your description, it should be changed to 
```
    PE_ABS_OP: begin
        add_a[0]    = ~op_a;
        add_b[0]    = 0;
        add_c_in[0] = 1'b1;
        res_p_w     = op_a[DATA_MSB];
        res_w       = diff_sign ? op_a : add_res[0];
        ovfl        = res_w[DATA_MSB];
    end
```
I thought that original code was your intension. I will change this rtl.

**Comment by tema8** [comment]
@kong0329 , 
The generator for developed primarily for the 3 input ALU and SAD operation rather that ABS. So the current code has a "lucky" copy-paste error that didn't changed main functionality, just the overflow.

Instead of :
        res_w       = diff_sign ? op_a : add_res[0];
It should have been:
        res_w       =~op_a[DATA_MSB] ? op_a : add_res[0];

Alternatively, you can keep the  res_w logic and change ovfl

**Comment by kong0329** [comment]
@leonardt 
To sum up,
1. `V` for signed `add` and `sub` -> no change. (align with Pat)
2. `V` for unsigned `add` and `sub` -> change to be same as current signed `V`
3. `abs` for both unsigned and signed -> a=0x8000

I am not clear with multiplication overflow. Can you clarify it?

**Comment by phanrahan** [comment]
How to set the overflow bit for multiplication is tricky, but I don’t remember the details.
In any event, I am quite sure it is different than V.

I would check the ARM / RISC5 arch manuals to see what they do.

**Comment by steveri** [comment]
@leonardt did this issue ever got sorted out to your satisfaction...?  I'm thinking we should not close this issue until we can point to where it is clearly spelled out in the architecture spec...

**Comment by leonardt** [comment]
I agree with leaving it open. I agree that it is a spec level issue that should be finalized before we can resolve it in the implementation/testing.

**Comment by steveri** [comment]
Huh. Still not fixed. When/if we have a definition for the flag(s), they should be part of the footnotes to the PE Flags table in the auto-generated spec:

https://github.com/StanfordAHA/CGRAGenerator/wiki/PE-Spec#pe_flags

Also: removing people who are probably no longer interested in this issue (i.e. everyone but me :)
